src/core/processing/qgsnativealgorithms.cpp

```diff
@@ -2618,37 +2618,62 @@ QgsRasterLayerUniqueValuesReportAlgorithm *QgsRasterLayerUniqueValuesReportAlgor
   return new QgsRasterLayerUniqueValuesReportAlgorithm();
 }
 
+bool QgsRasterLayerUniqueValuesReportAlgorithm::prepareAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback * )
+{
+  QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
+  mInterface.reset( layer->dataProvider()->clone() );
+  return true;
+}
+
 QVariantMap QgsRasterLayerUniqueValuesReportAlgorithm::processAlgorithm( const QVariantMap &parameters, QgsProcessingContext &context, QgsProcessingFeedback *feedback )
 {
   QgsRasterLayer *layer = parameterAsRasterLayer( parameters, QStringLiteral( "INPUT" ), context );
   int band = parameterAsInt( parameters, QStringLiteral( "BAND" ), context );
   QString outputFile = parameterAsFileOutput( parameters, QStringLiteral( "OUTPUT_HTML_FILE" ), context );
 
+  QHash< double, uint > uniqueValues;
+  uint layerWidth = layer->width();
+  uint layerHeight = layer->height();
+  uint layerSize = layerWidth * layerHeight;
 
-  QHash< double, int > uniqueValues;
-  int width = layer->width();
-  int height = layer->height();
-
-  QgsRasterBlock *rasterBlock = layer->dataProvider()->block( band, layer->extent(), width, height );
   int noDataCount = -1;
-  if ( rasterBlock->hasNoDataValue() )
+  if ( layer->dataProvider()->sourceHasNoDataValue( band ) )
```

Comment:
Also move this up to prepare, so that there's no need for "layer" here at all

Reply:
@nyalldawson , do I need to do this for layer->width() and layer->height() calls too?